### ISSUE TITLE
Support cloze deletions

### ISSUE THREAD

**Issue by parthopdas** [jasonwilliams/anki#7]
@jasonwilliams thanks for the wonderful tool.

would be good to have cloze deletion support.

**Comment by jasonwilliams** [comment]
So i could detect if the card title is using cloze syntax, then use the `cloze` card type if that's the case

**Comment by jasonwilliams** [comment]
Support added in https://github.com/jasonwilliams/anki/commit/cdae07c06fd6925b07bd05fcbe8b8d7390e338e9

**Comment by jasonwilliams** [comment]
@parthopdas could you please try 0.2.0 and let me know if that works as you expected? (only just published so may take time to propagate)

**Comment by lingocoder** [comment]
Thanks for adding _`cloze`_ support @jasonwilliams :+1: 

I've upgraded to 0.2.0. But I can't figure out how to make the extension create a _`cloze`_ card:

![anki vscode cloze bug 0 a](https://user-images.githubusercontent.com/49788472/90823465-d4b6ba80-e32d-11ea-8747-11dd58a5c3c5.gif)

> *i could detect if the card title is using cloze syntax, then use the `cloze` card type if that's the case*

I've tried:
1. Putting the card title in _`cloze`_ format
    * \# {{c1:Java SE}}
2. Putting the \#\# in _`cloze`_ format
    * \#\# {{c1:Anki for VSCode}}

When I look at the note type in the Anki browser, Anki says it's „*BasicWithHighlightVSCode*“. Of course, I could change it to the _`cloze`_ type manually there. But I was expecting the extension would do that for me.

What should I try next? TIA.

**Comment by jasonwilliams** [comment]
Strange!

Could you paste here the exact markdown text you're testing this with?
Cloze only works inside a H2 (##)

**Comment by lingocoder** [comment]
> *I'm assuming you know it should be 2 colons after the c1 right?*

Right. You can see in the screen recording, I do use double colons. I didn't escape them in my comment so, the editor rendered it as a single colon.

> *Could you paste here the exact markdown text you're testing this with?*

This is the markdown I first start with in the screen recording:

> \# {{c1::Java SE}}
>
> \#\# Version {{c2\:\:0.2.0}} will use the {{c3\:\:cloze}} card type if it detects the {{c4\:\:card title}} is using {{c3\:\:cloze}} syntax.
>
> %
>
> \[\*Support cloze deletions \| \#7\*\]\(https\://github.com/jasonwilliams/anki/issues/7\)

You can see in the recording, I also tried it the other way I listed in my previous comment.

> Cloze only works inside a H2 (\#\#). 

The markdown I pasted above has the `cloze`s in the H2.

> *i wouldn't be expecting many to be doing that with Cloze cards. What is your usecase?*

***TL;DR*** — The *back* sides of my _`cloze`s_ have come to be my own personal *Stack Overflow*. 

----

I have several dozen _`cloze`_ type cards. Every one of them has a *back* that has a couple sentences about whatever the _`cloze`_ is about. The backs of all my _`cloze`s_ also have a citation of where I got the review info from.

I find that approach very effective for the way *I* learn best. If I get a _`cloze`_ review wrong, having it show me a brief *reminder* is invaluable to me. 

The information on the back also serves as a kind of *reference*. I find myself sometimes opening Anki — *at random times of the day in between reviews* — for the sole purpose of looking up some information or URLS that I've noted on the *backs* of my _`clozes`_.

**Comment by jasonwilliams** [comment]
Did the removal of the % work for you?

**Comment by jasonwilliams** [comment]
Having a Cloze with a back is going to be tricky.
This is because the default cloze template only accepts 1 “Text” field, and making a new cloze template via the API isn’t really possible, so I’m not sure I can actually implement that.

The best I can do for now is allow extending the front.

**Comment by lingocoder** [comment]
I had already edited my original comment with answers to your questions, before I saw your other follow-up questions.

> *the default cloze template only accepts 1 “Text” field*

I have a this _`cloze`_ type that has the following two fields:

![anki fields for cloze 0](https://user-images.githubusercontent.com/49788472/90833215-77c40000-e33f-11ea-9ec4-fde9d2ce766d.png)

The „*Extra*“ field is what I referred to as „*the back*“ in my previous comments. 

I forget whether I created that _`cloze`_ note type myself, or whether I got it from an add-on or something. In any case, I use it more than any other note type:

![anki note types 0](https://user-images.githubusercontent.com/49788472/90833135-4a775200-e33f-11ea-8f79-42f0ebaf5a4d.png)

> *making a new cloze template via the API isn’t really possible*

I don't guess it's possible to have the extension use my pre-installed `_cloze`_ template? Is it?

**Comment by jasonwilliams** [comment]
You just have added Extra (somehow) as I don’t have that on mine.
Can you show your template for your Cloze card?

**Comment by lingocoder** [comment]
> *Haha no that’s not an option*

It's not uncommon for end users to able to *override* built-in properties of different software. I was just wondering if that sort of flexibility you're designed into your extension.

> *anything I do here needs to work for everyone*

Yeah. I think you misunderstood me. I was actually thinking of making that kind of extensibility available for *everyone*; not just me. 

By that, I mean there could be *some kind* of extensibility point where end users edit a configuration file of your extension, say. That way, ***anybody*** (*not just me*) can configure your extension to use whatever note type templates they want. *Customization* really is a thing; believe it or not. Haha.

> *Can you show your template for your Cloze card?*

![anki cloze deletion template 0](https://user-images.githubusercontent.com/49788472/90835487-a690a500-e344-11ea-8a05-c543487aadde.png)

**Comment by jasonwilliams** [comment]
> By that, I mean there could be some kind of extensibility point where end users edit a configuration file of your extension, say. That way, anybody (not just me) can configure your extension to use whatever note type templates they want. Customization really is a thing; believe it or not. Haha.

Yep I understand, this idea has floated around and for now it’s unlikely to happen any time soon.
It’s not as simple as “selecting a note to use and the extension sends to that”. Each note has specific fields and cards, so now you need to select what the fields are called and what type they are. This is the reason the extension uses its own note type, because that’s compatible with how the extension works.

Basically it gets messy, and I don’t think there’s much value in it at this point. If that’s needed it’s easier to change after the card has been sent.

**Comment by lingocoder** [comment]
> „*don’t think there’s much value in it at this point*“

The *idea* of your extension is appealing. But it's petty much strictly for people who use Anki's *Basic* note type exclusively.

I'm pretty good at HTML/CSS. So, in my current approach, I use an HTML editor to create my Anki cards. I just copy the markup from my editor and paste it into Anki.

Which is essentially what your extension does. Except your extension's *copy and paste* is done over TCP/HTTP.

> *If that’s needed it’s easier to change after the card has been sent*

It's easier to continue doing my copy and paste from my HTML editor like I've been doing; it's simpler and involves fewer steps. But thanks anyway.

Still a cool idea for a vscode extension though :+1:

**Comment by jasonwilliams** [comment]
> The idea of your extension is appealing. But it's petty much strictly for people who use Anki's Basic note type exclusively.

This extension is new, it was only created a couple of weeks ago. I have to start somewhere, and that somewhere is Basic but it’s the easiest to work with.

It can’t be as flexible as what you’re asking for because I’m limited by what the API can do, which is less than what you can do manually.

If I can’t create new Cloze templates then that means overwriting the templates that are already there, and I don’t think users will appreciate that.

It’s not like I don’t **want** to implement these things, some of them aren’t possible currently and maybe possible in future.

Feel free to raise feature requests on specific issues and I accept PRs
